File: linked_lists/linked_list.py
from node import Node

class LinkedList:

    def __init__(self, value):
        new_node = Node(value = value)
        self.head = new_node
        self.tail = new_node
        self.length = 1

    def is_empty(self):
        return not self.head and not self.tail

    def print_list(self):
        current = self.head
        while current:
            print(current.value)
            current = current.next

    def append(self, value):
        new_node = Node(value = value)
        if not self.is_empty():
            self.tail.next = new_node
            self.tail = new_node
        else:
            self.head = new_node
            self.tail = new_node
        self.length += 1
        return True

    def pop(self):

        if self.is_empty():
            return None

        if self.head is self.tail: # or self.length == 1
            tmp = self.head # or self.tail
            self.head = None
            self.tail = None
            self.length -= 1
            return tmp

        tmp = self.head
        while tmp.next is not self.tail:
            tmp = tmp.next
        self.tail = tmp
        self.tail.next = None
        self.length -= 1

        return tmp.next

    def prepend(self, value):
        new_node = Node(value = value)
        new_node.next = self.head
        if self.is_empty():
            self.tail = new_node
        self.head = new_node
        self.length += 1
        return True

    def pop_first(self):

        if self.is_empty():
            return None
        tmp = self.head
        self.head = self.head.next
        if self.head is self.tail:
            self.tail = None
        tmp.next = None
        self.length -= 1
        return tmp

    def get(self, idx):
        # Returns the node at the specified index
        if idx < 0 or idx > self.length - 1:
            return None

        if self.is_empty():
            return None

        i = 0
        node = self.head
        while i != idx:
            node = node.next
            i += 1

        return node

    def set_value(self, idx, value):
        # Sets a new value for the node with index idx
        node = self.get(idx = idx)
        if node:
            node.value = value
            return True
        return False

    def insert(self, idx, value):
        # Inserts a new node at a given index
        if idx < 0 or idx > self.length or self.is_empty():
            return False

        if idx == 0:
            self.prepend(value = value)
        elif idx == self.length:
            self.append(value = value)
        else:
            prev_node = self.get(idx = idx - 1)
            if prev_node:
                new_node = Node(value = value)
                new_node.next = prev_node.next
                prev_node.next = new_node
        self.length += 1
        return True

    def remove(self, idx):
        # Removes node at a given index
        node = None
        if idx < 0 or idx >= self.length or self.is_empty():
            return node
        if idx == 0:
            node = self.pop_first()
        elif idx == self.length - 1:
            node = self.pop()
        else:
            prev_node = self.get(idx = idx - 1)
            if prev_node:
                # tmp is the node to be removed
                tmp = prev_node.next
                node = tmp
                prev_node.next = tmp.next
                tmp.next = None
        self.length -= 1
        return node

    def reverse(self):
        if self.is_empty():
            return False

        tmp = self.head
        self.head = self.tail
        self.tail = tmp

        after = tmp.next
        before = None

        while not self.head.next:
            after = tmp.next
            tmp.next = before
            before = tmp
            tmp = after

File: linked_lists/linked_list.js
import Node from './node.js'

class LinkedList {
  constructor() {
    this.head = null; // the head is a reference to the first node in the linkedList
  }

  append(data) {
    const newNode = new Node(data); // create a new node to add to the list
    if (!this.head) {
      this.head = newNode; // if the list is empty, newNode becomes the head
    } else {
      let current = this.head; // start from the head

      // traverse to the last node by following the next pointers
      while (current.next) {
        current = current.next;
      }
      current.next = newNode; // set the new node as the last node's next pointer.
    }
  }

  // The display() method gathers all data from each node in the linked list and returns them as an array.
  display() {
    const elements = []; // array to hold all node data
    let current = this.head; // set up the head as the pointer
    while (current) { // loop until current is null (end of the list)
      elements.push(current.data); // add the current node's data to the array
      current = current.next; // move to the next node
    }
    return elements; // return the collected data as an array
  }

  // The find() method searches for a node containing a specific data value
  // if the node is found, it returns the node; otherwise, it returns null
  find(data) {
    let current = this.head; // start from the head
    while (current) {
      if (current.data === data) { // check if the current node contains the data
        return current; // return the Node if found
      }
      current = current.next
    }
    return null;
  }

  delete(data) {
    // check if the list is empty
    if (!this.head) return;

    // if the head contains the data, remove it by setting head to the next node
    if (this.head.data === data) {
      this.head = this.head.next; // remove the head node
    }

    // traverse the list to find the node just before the Target Node
    let current = this.head;
    while (current.next && current.next.data !== data) { // loop until the next node contains the target data or reaches the end
      current = current.next;
    }

    // if the target node is found, delete it by adjusting the next pointer
    if (current.next) {
      current.next = current.next.next; // skip the Target Node by linking to the node after it.
    }
  }
}

export default LinkedList;
